### ISSUE TITLE
Service fails to start without explanation on Windows Server 2012 R2 - both interactive and unattended (push-jobs)

### ISSUE THREAD

**Issue by DavidR91** [chef/opscode-pushy-client#109]
When installing on Windows Server 2012 R2, the installation 'ends prematurely' because the service cannot start.

If you do an unattended install (which prevents ruby.exe from being rolled back - for some reason), you can actually go into embedded/bin and run pushy-client.bat yourself - the output seems to imply it is functioning normally - suggesting the issue lies solely with the installer and/or the service wrapper?

## Push Jobs Client Version

Fresh installs of 2.1.3-1-x86

(talking to a 12.2.0 Chef Server with push jobs server 2.1.0)

## Platform Version

Windows Server 2012 R2

(both generic and the latest image in Azure)

## Replication Case
Install push client 2.1.3.1-x86 on Windows Server 2012 R2 (I've not found any commonalities - other than the chef server they are talking to - between machines exhibiting this problem. As stated above, this happens on clean Azure 2012 boxes)

## Stacktrace

(This is the output when executed by the push-jobs cookbook, but the 'cannot start in timely fashion' is presumably the same issue scuppering standadlone/interactive installs too)
```
Generated at 2016-11-10 15:17:32 +0000
Chef::Exceptions::MultipleFailures: Chef::Exceptions::MultipleFailures
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/ffi-win32-extensions-1.0.3/lib/ffi/win32/extensions.rb:64:in `raise_windows_error'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-service-0.8.9/lib/win32/service.rb:736:in `start'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:102:in `block in start_service'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:288:in `block in spawn_command_thread'

>>>> Caused by SystemCallError: windows_service[push-jobs-client] (push-jobs::service_windows line 63) had an error: SystemCallError: The service did not respond to the start or control request in a timely fashion. - StartService: The service did not respond to the start or control request in a timely fashion.
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/ffi-win32-extensions-1.0.3/lib/ffi/win32/extensions.rb:64:in `raise_windows_error'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-service-0.8.9/lib/win32/service.rb:736:in `start'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:102:in `block in start_service'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:288:in `block in spawn_command_thread'

>>>> Caused by SystemCallError: The service did not respond to the start or control request in a timely fashion. - StartService: The service did not respond to the start or control request in a timely fashion.
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/ffi-win32-extensions-1.0.3/lib/ffi/win32/extensions.rb:64:in `raise_windows_error'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/win32-service-0.8.9/lib/win32/service.rb:736:in `start'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:102:in `block in start_service'
C:/opscode/chef/embedded/lib/ruby/gems/2.3.0/gems/chef-12.16.42-universal-mingw32/lib/chef/provider/service/windows.rb:288:in `block in spawn_command_t
```

The only event viewer messages are

```
A service was installed in the system.

Service Name:  Push Jobs Client Service
Service File Name:  c:\opscode\push-jobs-client\embedded\bin\ruby.exe c:\opscode\push-jobs-client\embedded\lib\ruby\gems\2.1.0\gems\opscode-pushy-client-2.1.3\lib\pushy_client\windows_service.rb
Service Type:  user mode service
Service Start Type:  auto start
Service Account:  LocalSystem
```

`A timeout was reached (30000 milliseconds) while waiting for the Push Jobs Client Service service to connect.`

I cannot find any other logs pertaining to push-client for **why** it would have failed to start

## Related issues
I have seen #63 and it could well be a cause - the issue is I don't know how to determine if/whether the client.rb is deemed 'incorrect'

**Comment by rhass** [comment]
This may be related to issue #63.

**Comment by DavidR91** [comment]
I think I've managed to squeeze out some more useful detail:

The `ruby.exe` command line which was being executed is 
`C:\opscode\push-jobs-client\embedded\bin\ruby.exe C:\opscode\push-jobs-client\embedded\lib\ruby\gems\2.1.0\gems\opscode-pushy-client-2.1.3\lib\pushy_client\windows_service.rb`

Which is generating:
```
C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:2285:in `raise_if_conflicts': Unable to activate opscode-pushy-client-2.1.3, because ohai-8.21.0 conflicts with ohai (= 8.19.0) (Gem::ConflictError)
        from C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems/specification.rb:1408:in `activate'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems.rb:196:in `rescue in try_activate'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems.rb:193:in `try_activate'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:125:in `rescue in require'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/site_ruby/2.1.0/rubygems/core_ext/kernel_require.rb:40:in `require'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/gems/2.1.0/gems/opscode-pushy-client-2.1.3/lib/pushy_client/job_runner.rb:24:in `<top (required)>'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/gems/2.1.0/gems/opscode-pushy-client-2.1.3/lib/pushy_client.rb:20:in `require_relative'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/gems/2.1.0/gems/opscode-pushy-client-2.1.3/lib/pushy_client.rb:20:in `<top (required)>'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/gems/2.1.0/gems/opscode-pushy-client-2.1.3/lib/pushy_client/windows_service.rb:25:in `require_relative'
        from C:/opscode/push-jobs-client/embedded/lib/ruby/gems/2.1.0/gems/opscode-pushy-client-2.1.3/lib/pushy_client/windows_service.rb:25:in `<main>'
```

after extracting out the push-jobs-client.zip from the MSI, it appears both of those ohai gems are installed by default (e.g. one of them did not already exist on my machine or anything like that)

`embedded\lib\ruby\gems\2.1.0\gems\ohai-8.19.0`
and 
`embedded\lib\ruby\gems\2.1.0\gems\ohai-8.21.0`

So... is this just broken out of the box, or is there something in my environment that is prompting ruby to activate one of these etc.?

**Comment by DavidR91** [comment]
After some experimentation, removing this condition:

https://github.com/chef/opscode-pushy-client/blob/master/lib/pushy_client/job_runner.rb#L23

```
if Chef::Platform.windows?
  require 'pushy_client/win32'
end
```

resolves my issue and the service can start and operate correctly.

Is it possible that this require is no longer required, e.g. with this particular version of Ruby, or something?

**Comment by ianau74** [comment]
I had the same issue.  I just got rid of this directory and the service started up fine:
C:\opscode\push-jobs-client\embedded\lib\ruby\gems\2.1.0\gems\ohai-8.21.0

**Comment by DavidR91** [comment]
Any idea how to modify the installer not to include the gem in the first place? (i.e. as a pull request?)

I can't see any neat way I can do that workaround if I'm installing the client automatically (since the MSI will still fail before any manual file deletions kick in - e.g. if I do it in a cookbook)

**Comment by rhass** [comment]
@DavidR91 We are have an internal backlog item as result of this bug report to fix this (thank you by the way!). However, I don't personally know which release it is targeted to fixed. With the holidays upon us the I can say is :soon:

**Comment by DavidR91** [comment]
@rhass that's absolutely fine - just happy to see it acknowledged (also thanks @ianau74 for confirming this was not just my environment + providing a simpler workaround)

**Comment by DavidR91** [comment]
Confirmed fixed as of 2.1.4

**Comment by charlesjohnson** [comment]
Great news! Going to close this out as resolved.
